hivemind/p2p/p2p_daemon.py

```diff
@@ -146,27 +150,41 @@ async def send_raw_data(byte_str, writer):
         writer.write(request)
 
     @staticmethod
-    async def send_data(data, writer):
+    async def send_message_pack(data, writer):
         raw_data = MSGPackSerializer.dumps(data)
         await P2P.send_raw_data(raw_data, writer)
 
     @staticmethod
     async def send_protobuf(protobuf, out_proto_type, writer):
         if type(protobuf) != out_proto_type:
-            error = TypeError('Unary handler returned protobuf of wrong type.')
-            await P2P.send_raw_data(MSGPackSerializer.dumps(error), writer)
-            raise error
+            raise TypeError('Unary handler returned protobuf of wrong type.')
         await P2P.send_raw_data(protobuf.SerializeToString(), writer)
 
     @staticmethod
-    async def receive_raw_data(reader: asyncio.StreamReader):
-        header = await reader.readexactly(P2P.HEADER_LEN)
+    async def send_protobuf_with_error(protobuf, out_proto_type, writer):
+        if type(protobuf) != out_proto_type:
+            raise TypeError('Unary handler returned protobuf of wrong type.')
+        if out_proto_type == p2pd_pb2.P2PRPCError:
+            await P2P.send_raw_data(P2P.ERROR_MESSAGE, writer)
+        else:
+            await P2P.send_raw_data(P2P.RESULT_MESSAGE, writer)
```

Comment:
```suggestion
            writer.write(P2P.ERROR_MESSAGE)
        else:
            writer.write(P2P.RESULT_MESSAGE)
```

I'd suggest to send one byte instead of 9 bytes (`HEADER_LEN` + the type byte) here.

Reply:
From my point of view it's better to keep send_raw_data as the only one place where `writer.write` is called.